src/backends/perf_pt/test_helpers.c

```diff
@@ -0,0 +1,144 @@
+// freely licensable by each licensor hereunder covering either (i) the
+// unmodified Software as contributed to or provided by such licensor, or (ii)
+// the Larger Works (as defined below), to deal in both
+//
+// (a) the Software, and
+// (b) any piece of software and/or hardware listed in the lrgrwrks.txt file
+// if one is included with the Software (each a "Larger Work" to which the Software
+// is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#define _GNU_SOURCE
+
+#include <link.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+#include <inttypes.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/auxv.h>
+#include <err.h>
+#include <elf.h>
+
+#include "perf_pt_private.h"
+
+// Arguments for perf_pt_append_self_ptxed_raw_args().
+struct append_args {
+    int vdso_fd;
+    char *vdso_filename;
+    void *rs_args_vec;
+};
+
+// Public prototypes.
+bool perf_pt_append_self_ptxed_raw_args(struct append_args *);
+
+// Private prototypes.
+static int append_self_ptxed_raw_args_cb(struct dl_phdr_info *, size_t, void *);
+
+// Rust functions.
+extern void push_ptxed_arg(void *, const char *);
+
+// More than enough space for a ptxed `--raw` argument parameter.
+#define MAX_RAW_ARG 128
+
+/*
+ * Walks the program headers pushing ptxed `--raw` arguments for the current
+ * process into the supplied (pointer to a) Rust `AppendSelfPtxedArgs`
+ * instance.
+ */
+bool
+perf_pt_append_self_ptxed_raw_args(struct append_args *rs_args)
+{
+    if (dl_iterate_phdr(append_self_ptxed_raw_args_cb, rs_args) == 1) {
+        return false;
+    }
+    return true;
+}
+
+/*
+ * Callback for perf_pt_append_self_ptxed_raw_args.
+ *
+ * `data` is a pointer to a Rust `AppendSelfPtxedArgs` instance.
+ */
+static int
+append_self_ptxed_raw_args_cb(struct dl_phdr_info *info, size_t size, void *data)
```

Comment:
Most of this function is the same as `load_self_image_cb` -- can/should we factor out the common bits?

Reply:
I did consider it, but thought it might be simpler to comprehend by having them separate like this. How strongly do you feel about it?

It might be a bit weird/confusing to have `void *data` be either a pointer to a Rust struct or a C `pt_image`.

Reply:
It's always a judgement call as to when to copy/paste and when to factor out. Your decision (I just wanted to flag it).

Reply:
I think in this case it would be best to leave them apart for the reason above.

Reply:
OK.